# HMS Commander Examples: Overview and Learning Path

This notebook provides an overview of the hms-commander example notebooks and a recommended learning path from beginner to advanced topics.

**hms-commander** is a Python library for automating HEC-HMS (Hydrologic Engineering Center's Hydrologic Modeling System) operations, following the architectural patterns established by ras-commander.

In [1]:
# pip install hms-commander

**For Development**: If working on hms-commander source code, use the `hmscmdr_local` conda environment (editable install) instead of pip install. See the development environment guide in `.claude/rules/project/development-environment.md`.

## Environment Verification

Run this cell to verify your installation is working correctly:

In [2]:
# Verify installation and detect HMS versions
import sys
print(f"Python: {sys.version}")
print(f"Python executable: {sys.executable}")
print()

# Check hms-commander installation
try:
    import hms_commander
    print(f"hms-commander: INSTALLED")
    print(f"  Location: {hms_commander.__file__}")
except ImportError:
    print("hms-commander: NOT INSTALLED - run 'pip install hms-commander'")

print()

# Check HMS installations
try:
    from hms_commander import HmsExamples
    versions = HmsExamples.list_versions()
    if versions:
        print(f"HEC-HMS installations detected: {len(versions)}")
        for v in versions:
            print(f"  - HMS {v}")
    else:
        print("HEC-HMS: No installations detected")
        print("  Note: Notebooks requiring HMS execution will not work without HMS installed.")
except Exception as e:
    print(f"HEC-HMS detection: Error - {e}")

print()

# Check optional dependencies
print("Optional dependencies:")
for pkg in ['pyjnius', 'geopandas', 'xarray', 'matplotlib']:
    try:
        __import__(pkg)
        print(f"  {pkg}: INSTALLED")
    except ImportError:
        print(f"  {pkg}: not installed (optional)")

print()
print("Verification complete!")

Python: 3.13.5 | packaged by Anaconda, Inc. | (main, Jun 12 2025, 16:37:03) [MSC v.1929 64 bit (AMD64)]
Python executable: C:\Users\billk_clb\anaconda3\python.exe



2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.10 at C:\Program Files\HEC\HEC-HMS\4.10


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.11 at C:\Program Files\HEC\HEC-HMS\4.11


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.12 at C:\Program Files\HEC\HEC-HMS\4.12


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.13 at C:\Program Files\HEC\HEC-HMS\4.13


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.4.1 at C:\Program Files\HEC\HEC-HMS\4.4.1


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.5 at C:\Program Files\HEC\HEC-HMS\4.5


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.6 at C:\Program Files\HEC\HEC-HMS\4.6


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.7.1 at C:\Program Files\HEC\HEC-HMS\4.7.1


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.8 at C:\Program Files\HEC\HEC-HMS\4.8


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.9 at C:\Program Files\HEC\HEC-HMS\4.9


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.0.0 at C:\Program Files (x86)\HEC\HEC-HMS\3.0.0


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.0.1 at C:\Program Files (x86)\HEC\HEC-HMS\3.0.1


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.1.0 at C:\Program Files (x86)\HEC\HEC-HMS\3.1.0


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.2 at C:\Program Files (x86)\HEC\HEC-HMS\3.2


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.3 at C:\Program Files (x86)\HEC\HEC-HMS\3.3


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.4 at C:\Program Files (x86)\HEC\HEC-HMS\3.4


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 3.5 at C:\Program Files (x86)\HEC\HEC-HMS\3.5


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.0 at C:\Program Files (x86)\HEC\HEC-HMS\4.0


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.1 at C:\Program Files (x86)\HEC\HEC-HMS\4.1


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.2.1 at C:\Program Files (x86)\HEC\HEC-HMS\4.2.1


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found HMS 4.3 at C:\Program Files (x86)\HEC\HEC-HMS\4.3


2026-01-08 13:11:16 - hms_commander.HmsExamples - INFO - Found 21 HMS installation(s) with examples


hms-commander: INSTALLED
  Location: C:\Users\billk_clb\anaconda3\Lib\site-packages\hms_commander\__init__.py

HEC-HMS installations detected: 21
  - HMS 4.13
  - HMS 4.12
  - HMS 4.11
  - HMS 4.10
  - HMS 4.9
  - HMS 4.8
  - HMS 4.7.1
  - HMS 4.6
  - HMS 4.5
  - HMS 4.4.1
  - HMS 4.3
  - HMS 4.2.1
  - HMS 4.1
  - HMS 4.0
  - HMS 3.5
  - HMS 3.4
  - HMS 3.3
  - HMS 3.2
  - HMS 3.1.0
  - HMS 3.0.1
  - HMS 3.0.0

Optional dependencies:
  pyjnius: not installed (optional)
  geopandas: INSTALLED


  xarray: INSTALLED


  matplotlib: INSTALLED

Verification complete!


## HMS Terminology Glossary

If you're new to HEC-HMS, here are key terms used throughout these notebooks:

| Term | Description |
|------|-------------|
| **Subbasin** | A drainage area that contributes runoff to a single outlet point |
| **Reach** | A channel segment that routes flow downstream |
| **Junction** | A point where multiple flows combine (confluence) |
| **Reservoir** | A storage element that attenuates peak flows |
| **Basin Model** | The physical watershed representation (.basin file) |
| **Met Model** | Precipitation and evapotranspiration inputs (.met file) |
| **Control Specs** | Time window and time step for simulation (.control file) |
| **Run** | A combination of Basin + Met + Control specifications |
| **DSS File** | HEC Data Storage System file containing time series results |
| **Hyetograph** | Time distribution of precipitation depths |
| **Hydrograph** | Time series of flow rates at a point |

## Learning Path

The notebooks are organized into progressive tracks:

### Beginner Track (Start Here) - Estimated Time: 30-45 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 00 | Overview and Learning Path | This notebook - orientation and prerequisites | 5 min |
| 01 | Basic HMS Workflow | Initialize project, execute run, extract results | 10-15 min |
| 02 | Project DataFrames | Explore project structure via DataFrames | 10-15 min |

### Intermediate Track - Estimated Time: 60-90 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 03 | File Operations | Work with HmsBasin, HmsMet, HmsControl, HmsGage | 15-20 min |
| 04 | Run Management | Configure and validate simulation runs | 10-15 min |
| 05 | Clone Workflow | Non-destructive model modifications for QAQC | 20-30 min |
| 06 | Results and DSS | HmsDss and HmsResults for data extraction | 15-20 min |

### Advanced Track - Estimated Time: 45-90 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 07 | Execution and Jython | Version detection, script generation, batch execution | 15-20 min |
| 08 | M3 Models | HCFCD M3 model discovery and extraction | 15-30 min |
| 09 | M3 Conversion | HMS 3.x to 4.x project conversion workflow | 30-60 min |

### Storm Generation Track - Estimated Time: 30-45 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 10 | Atlas 14 Hyetograph | Generate design storms from NOAA Atlas 14 | 15-20 min |
| 11 | Frequency Storm | Variable duration storms using TP-40/Hydro-35 patterns | 15-20 min |

### Validation and Equivalence Proofs - Estimated Time: 20-30 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 12 | SCS Type Validation | Proof of equivalence: ScsTypeStorm vs HEC-HMS | 10-15 min |
| 13 | Atlas 14 Multi-Duration Validation | Proof of equivalence: Atlas14Storm across durations | 10-15 min |

### Integration Track (AORC Gridded Precipitation) - Estimated Time: 30-60 minutes

| Notebook | Title | Description | Time |
|----------|-------|-------------|------|
| 14a | AORC Download | Download AORC precipitation data | 10-20 min |
| 14b | AORC Grid Setup | Create grid definitions and HRAP cell mappings | 10-15 min |
| 14c | AORC HMS Execution | Run HMS with gridded precipitation and analyze results | 15-20 min |

## Prerequisites and Data Requirements

### All Notebooks

- Python 3.10+
- `pip install hms-commander`
- HEC-HMS 4.x installation (for execution examples)

### Example Projects (Automatic)

Most notebooks use `HmsExamples.extract_project()` which automatically extracts example projects from your HEC-HMS installation. No manual downloads needed.

```python
from hms_commander import HmsExamples

# Extracts 'tifton' from your HMS installation
project_path = HmsExamples.extract_project("tifton")
```

### M3 Models (Notebooks 08-09)

**Requires M3 data download.** M3 models are HCFCD FEMA-effective H&H models that are downloaded from the HCFCD model library.

```python
from hms_commander import HmsM3Model

# First run downloads from HCFCD (may take time)
project_path = HmsM3Model.extract_project('D', 'D100-00-00')
```

### AORC Data (Notebooks 14a-c)

**Requires AORC data download.** AORC (Analysis of Record for Calibration) data is downloaded from NOAA AWS.

```python
from hms_commander import HmsAorc

# Downloads from NOAA AWS S3 (may take time for large areas)
aorc_data = HmsAorc.download(bounds, start_time, end_time, output_path)
```

### DSS Operations (Notebooks 01, 06)

DSS file reading requires:
- Java 8+ JDK/JRE
- `pip install pyjnius`

The HEC Monolith libraries are auto-downloaded on first use.

## HMS Classes by Notebook

| Class | Primary Notebook | Description |
|-------|------------------|-------------|
| `init_hms_project` | 01, 02 | Initialize HMS project and global `hms` object |
| `HmsPrj` | 02 | Project dataframes and component discovery |
| `HmsExamples` | 01, 02 | Extract example projects from HMS installations |
| `HmsCmdr` | 01, 07 | High-level simulation execution |
| `HmsResults` | 01, 06 | Extract peak flows, hydrographs, volumes |
| `HmsDss` | 06 | DSS file operations (read/write) |
| `HmsBasin` | 03, 05 | Basin model file operations |
| `HmsMet` | 03 | Meteorologic model operations |
| `HmsControl` | 03 | Control specification operations |
| `HmsGage` | 03 | Time series gage operations |
| `HmsRun` | 04 | Run configuration and validation |
| `HmsJython` | 07 | Jython script generation |
| `HmsM3Model` | 08, 09 | HCFCD M3 model access |
| `Atlas14Storm` | 10, 13 | NOAA Atlas 14 hyetograph generation |
| `FrequencyStorm` | 11 | TP-40/Hydro-35 frequency storms |
| `ScsTypeStorm` | 12 | SCS Type I, IA, II, III storms |
| `HmsHuc` | 14a | HUC watershed boundary download |
| `HmsAorc` | 14a | AORC precipitation download |
| `HmsGrid` | 14b | Grid definition and cell mapping |

## Validation and Equivalence Proof Notebooks

Notebooks 12 and 13 are **validation and equivalence proofs** that demonstrate hms-commander's storm generation algorithms produce identical results to HEC-HMS.

These notebooks:
- Compare hms-commander output to HEC-HMS ground truth
- Validate to 10^-6 precision
- Serve as certification for production use

**For users**: These prove the algorithms are correct.
**For developers**: These serve as regression tests.

## Next Steps

1. **New to hms-commander?** Start with **01_basic_workflow.ipynb**
2. **Want to explore project structure?** Continue to **02_project_dataframes.ipynb**
3. **Need file operations?** See **03_file_ops_basin_met_control_gage.ipynb**
4. **Working with M3 models?** Jump to **08_m3_models.ipynb**
5. **Generating design storms?** See **10_atlas14_hyetograph.ipynb**

Each notebook is self-contained and includes setup, examples, and validation checks.